### PR TITLE
Create a flutter example application that uses multiple client implementations

### PR DESCRIPTION
**Pull request by brianquinlan**
* Creates an example flutter application (in the style of the flutter gallery) that demonstrates `package:http` including:
  * how to do configuration (including web configuration)
  * how to pass `Client` instances through the app
  * simple usage 
* Updates the `cronet_http` and `cupertino_http` examples to match the new application, except where this would increase the complexity too much.
* The application was created with `flutter create -t app`. Some interesting files:
  * README.md
  * pkgs/cronet_http/example/lib/main.dart
  * pkgs/cronet_http/example/lib/main.dart
  * pkgs/flutter_http_example/README.md
  * pkgs/flutter_http_example/lib/*
  * pkgs/flutter_http_example/mono_pkg.yaml
  * pkgs/flutter_http_example/pubspec.yaml
  * pkgs/flutter_http_example/test/widget_test.dart
---

- [X] I’ve reviewed the contributor guide and applied the relevant portions to this PR.

<details>
  <summary>Contribution guidelines:</summary><br>

- See our [contributor guide](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md) for general expectations for PRs.
- Larger or significant changes should be discussed in an issue before creating a PR.
- Contributions to our repos should follow the [Dart style guide](https://dart.dev/guides/language/effective-dart) and use `dart format`.
- Most changes should add an entry to the changelog and may need to [rev the pubspec package version](https://github.com/dart-lang/sdk/wiki/External-Package-Maintenance#making-a-change).
- Changes to packages require [corresponding tests](https://github.com/dart-lang/.github/blob/main/CONTRIBUTING.md#Testing).

Note that many Dart repos have a weekly cadence for reviewing PRs - please allow for some latency before initial review feedback.
</details>

### CODE REVIEW

File: pkgs/cronet_http/example/lib/main.dart
Comment (natebosch) [inline pkgs/cronet_http/example/lib/main.dart:84]: [optional] consider flipping the condition and using an early return guard clause

```suggestion
    // Avoid the situation where a slow-running query finishes late and
    // replaces newer search results.
    if (query != _lastQuery) return;
    setState(() {
      _books = books;
    });
```

File: pkgs/cupertino_http/example/lib/main.dart
Comment (natebosch) [inline pkgs/cupertino_http/example/lib/main.dart:82]:
```suggestion
    // Avoid the situation where a slow-running query finishes late and
    // replaces newer search results.
    if (query == _lastQuery) return;
    setState(() {
     _books = books;
    });
```

File: pkgs/flutter_http_example/README.md
Comment (natebosch) [inline pkgs/flutter_http_example/README.md:43]: [optional] WDYT about replacing "file" with "library".

```suggestion
This library used to create `package:http` `Client`s when the app is run inside
the Dart virtual machine, meaning all platforms except the web browser.

### `http_client_factory_web.dart`

This library used to create `package:http` `Client`s when the app is run inside
a web browser.

Web configuration must be done in a seperate library because Dart code cannot
import `dart:ffi` or `dart:io` when run in a web browser.

### `main.dart`

This library demonstrates how to:

* import `http_client_factory.dart` or `http_client_factory_web.dart`,
  depending on whether we are targeting the web browser or not.
* share a `package:http` `Client` by using `runWithClient` and
  `package:provider`.
* call `package:http` functions.

### `widget_test.dart`

This library demonstrates how to construct tests using `MockClient`.
```

File: pkgs/flutter_http_example/ios/Runner/Assets.xcassets/LaunchImage.imageset/README.md
Comment (natebosch) [inline pkgs/flutter_http_example/ios/Runner/Assets.xcassets/LaunchImage.imageset/README.md:3]: Is it typical flutter practice to check in these generated files?

File: pkgs/flutter_http_example/lib/book.dart
Comment (natebosch) [inline pkgs/flutter_http_example/lib/book.dart:1]:
```suggestion
// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
```

File: pkgs/flutter_http_example/lib/book.dart
Comment (natebosch) [inline pkgs/flutter_http_example/lib/book.dart:16]: [optional]

```suggestion
    if (json['items'] case final List jsonItems) {
      final items = jsonItems.cast<Map<String, Object?>>();
```

File: pkgs/flutter_http_example/lib/book.dart
Comment (natebosch) [inline pkgs/flutter_http_example/lib/book.dart:20]: [optional]
```suggestion
        if (item case {'volumeInfo': final volumeInfo as Map}) {
```

File: pkgs/flutter_http_example/lib/book.dart
Comment (natebosch) [inline pkgs/flutter_http_example/lib/book.dart:29]: [optional]

```suggestion
          if (volumeInfo
              case {
                'title': final String title,
                'description': final String description,
                'imageLinks': {'smallThumbnail': final String thumbnail}
              }){
            books.add(Book(title, description, thumbnail));
          }
```

File: pkgs/flutter_http_example/lib/main.dart
Comment (natebosch) [inline pkgs/flutter_http_example/lib/main.dart:1]:
```suggestion
// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
```

**Review by natebosch — APPROVED** [review]
I did not look at stuff other than the Dart. Consider getting a second review from someone with Flutter experience.

File: pkgs/flutter_http_example/README.md
Comment (brianquinlan) [reply pkgs/flutter_http_example/README.md:43]: Is a single dart file a library?

File: pkgs/flutter_http_example/ios/Runner/Assets.xcassets/LaunchImage.imageset/README.md
Comment (brianquinlan) [reply pkgs/flutter_http_example/ios/Runner/Assets.xcassets/LaunchImage.imageset/README.md:3]: It is checked in in the flutter gallery:
https://github.com/flutter/gallery/blob/fa031bfe9d131010e7a56ee5d343f9f85b367d64/ios/Runner/Assets.xcassets/LaunchImage.imageset/README.md?plain=1#L3

File: pkgs/flutter_http_example/lib/main.dart
Comment (brianquinlan) [reply pkgs/flutter_http_example/lib/main.dart:1]: This is copied from pkgs/cupertino_http so I think that this copyright is correct.

File: pkgs/flutter_http_example/README.md
Comment (natebosch) [reply pkgs/flutter_http_example/README.md:43]: A single import is a library. Many of our libraries are 1:1 with files, but part files allow a library to span more that one file.

**Pull request merged**
